journal/src/main/java/io/camunda/zeebe/journal/file/SegmentLoader.java

```diff
@@ -10,39 +10,46 @@
 import io.camunda.zeebe.journal.JournalException;
 import io.camunda.zeebe.journal.file.record.CorruptedLogException;
 import io.camunda.zeebe.util.FileUtil;
-import java.io.File;
 import java.io.IOException;
 import java.nio.ByteBuffer;
 import java.nio.ByteOrder;
 import java.nio.MappedByteBuffer;
 import java.nio.channels.FileChannel;
 import java.nio.channels.FileChannel.MapMode;
 import java.nio.file.FileAlreadyExistsException;
-import java.nio.file.OpenOption;
+import java.nio.file.Files;
+import java.nio.file.Path;
 import java.nio.file.StandardOpenOption;
-import java.util.Collections;
-import java.util.HashSet;
-import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /** Create a segment file. Load a segment from the segment file. */
 final class SegmentLoader {
-
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentLoader.class);
   private static final ByteOrder ENDIANNESS = ByteOrder.LITTLE_ENDIAN;
-  private final long lastWrittenIndex;
-  private final JournalIndex journalIndex;
 
-  SegmentLoader(final long lastWrittenIndex, final JournalIndex journalIndex) {
-    this.lastWrittenIndex = lastWrittenIndex;
-    this.journalIndex = journalIndex;
+  private final boolean preallocateFiles;
+
+  SegmentLoader() {
+    this(true);
   }
 
-  JournalSegment createSegment(final JournalSegmentDescriptor descriptor, final File segmentFile) {
+  SegmentLoader(final boolean preallocateFiles) {
+    this.preallocateFiles = preallocateFiles;
+  }
+
+  JournalSegment createSegment(
+      final Path segmentFile,
+      final JournalSegmentDescriptor descriptor,
+      final long lastWrittenIndex,
+      final JournalIndex journalIndex) {
     final MappedByteBuffer mappedSegment;
 
     try {
-      mappedSegment = mapNewSegment(segmentFile, descriptor);
+      mappedSegment = mapNewSegment(segmentFile, descriptor, lastWrittenIndex);
     } catch (final IOException e) {
-      throw new JournalException(String.format("Failed to map new segment %s", segmentFile), e);
+      throw new JournalException(
+          String.format("Failed to allocate new segment file %s", segmentFile), e);
```

Comment:
:wrench: I suppose it'd be evident from the stacktrace but the message could be adjusted depending on `preallocateFiles`

Reply:
Any suggestions? The only thing I can imagine is the word `allocate` might alarm users who've turned off pre-allocation. So we could reword this to, say, create. That's fairly neutral, and as you said, the cause will give you the correct error depending on whether you're preallocating or not.

Or did you mean that users would want it explicit here whether they are pre-allocating or not?

Reply:
I think we are on the same page here :+1: